java/sage/MediaServerRemuxer.java

```diff
@@ -0,0 +1,1008 @@
+      {
+        try
+        {
+          switchLock.wait(500);
+        }
+        catch (InterruptedException e)
+        {
+          break;
+        }
+
+        if (timeout-- <= 0)
+        {
+          if (Sage.DBG)
+            System.out.println("WARNING Could not find transition point after over 30 seconds!");
+          doSwitch(true);
+          break;
+        }
+      }
+
+      return !switching;
+    }
+  }
+
+  private void doSwitch()
+  {
+    doSwitch(false);
+  }
+
+  private void doSwitch(boolean force)
+  {
+    int searchBytes = writeBuffer.position();
+
+    int switchIndex = getSwitchIndex(writeBuffer, 0, searchBytes);
+    switchData += searchBytes;
+
+    if (switchIndex == -1)
+    {
+      if (switchData > SWITCH_BYTES_LIMIT)
+      {
+        if (Sage.DBG) System.out.println(
+            "WARNING Could not find transition point after searching 8 MB of data in stream!");
+
+        switchIndex = 0;
+      }
+      else if (force)
+      {
+        switchIndex = 0;
+      }
+    }
+
+    try
+    {
+      if (switchIndex != -1)
+      {
+        synchronized (switchLock)
+        {
+          if (currentFile != null)
+            remuxerMap.remove(currentFile);
+
+          if (switchIndex != 0)
+          {
+            int oldLimit = writeBuffer.position();
+
+            writeBuffer.position(switchIndex);
+            writer.writeFile(writeBuffer);
+
+            writeBuffer.limit(oldLimit).position(switchIndex);
+            writeBuffer.compact();
+          }
+
+          mediaServer.closeFile(false);
+          mediaServer.openWriteFile(switchFilename, switchUploadId, false);
+
+          this.fileChannel = mediaServer.getFileChannel();
+
+          // The thread switching the file is the same thread that writes to the file, so this will
+          // not create a race condition.
+          fileSize.set(0);
+
+          writer.writeFile(writeBuffer);
+
+          currentFile = new File(switchFilename);
+          if (interAssist)
+            remuxerMap.put(currentFile, this);
+
+          switching = false;
+          switchLock.notifyAll();
+        }
+      }
+      else
+      {
+        // Write out the current buffer contents so we don't check the same data twice.
+        writer.writeFile(writeBuffer);
+      }
+    }
+    catch (IOException e)
+    {
+      if (Sage.DBG)
+      {
+        System.out.println("ERROR in MediaServerRemuxer while switching:" + e);
+        e.printStackTrace(System.out);
+      }
+    }
+  }
+
+  private int getSwitchIndex(ByteBuffer data, int offset, int length)
+  {
+    // H.264 and MPEG-2 are the only formats supported.
+    if (!h264 && !mpeg2)
+    {
+      if (Sage.DBG)
+        System.out.println("CANNOT perform fast switch on a non-H264/MPEG2 MPEG2 TS file!");
+      return 0;
+    }
+
+    if (outputFormat == MPEGParser2.REMUX_TS)
+    {
+      if (videoPid == -1)
+      {
+        if (Sage.DBG) System.out.println("CANNOT perform fast switch without a video PID!");
```

Comment:
I reworded those. I borrowed some text from another Java class for the wording there. I've been trying to keep things reasonably consistent. :)